### ISSUE TITLE
DeepSHAP baseline x,y do not take mean values?

### ISSUE THREAD

**Issue by rightx2** [shap/shap#1159]
As I know of, when calculating slope, it needs a baseline x and baseline y and the baseline means taking a mean value of dataset according to https://towardsdatascience.com/interpretable-neural-networks-45ac8aa91411)

However, there is no code of "taking mean" in `nonlinear_1d` in `PytorchExplainer`

```
def nonlinear_1d(module, grad_input, grad_output):
    delta_out = module.y[: int(module.y.shape[0] / 2)] - module.y[int(module.y.shape[0] / 2):]
    delta_in = module.x[: int(module.x.shape[0] / 2)] - module.x[int(module.x.shape[0] / 2):]
    dup0 = [2] + [1 for i in delta_in.shape[1:]]
    # handles numerical instabilities where delta_in is very small by
    # just taking the gradient in those cases
    grads = [None for _ in grad_input]
    grads[0] = torch.where(torch.abs(delta_in.repeat(dup0)) < 1e-6, grad_input[0],
                           grad_output[0] * (delta_out / delta_in).repeat(dup0))
    return tuple(grads)
```

Shouldn't it be something like below, taking mean of `self.data`'s y to generate baseline?
```
delta_out = module.y[: int(module.y.shape[0] / 2)] - module.y[int(module.y.shape[0] / 2):].mean()
```

**Comment by rightx2** [comment]
@gabrieltseng

**Comment by gabrieltseng** [comment]
Hi @rightx2 - the deeplift gradients are individually calculated for each instance in the background.

The mean is then calculated on the final gradients:
https://github.com/slundberg/shap/blob/master/shap/explainers/deep/deep_pytorch.py#L198

**Comment by rightx2** [comment]
@gabrieltseng I see. I missed that point. Thanks. 

I have a question. I tried not to use `repeat()` operation on https://github.com/slundberg/shap/blob/93dc2a1e446616fb0858b2ec108f80e4969ba6d9/shap/explainers/deep/deep_pytorch.py#L179 because If so, it will speed up the `autograd()` a lot. 

Unfortunately, I tried some corrections but failed to pass the unit test. Do you think it is possible to do or technically not?
